mmeds/util.py

```diff
@@ -1383,19 +1384,199 @@ def run_analysis(path, tool_type, testing=False):
         raise e
 
 
-def upload_study_directly(queue, study_name, subject, subject_type, specimen, user):
+def upload_study_local(queue, study_name, subject, subject_type, specimen, user):
     """
-    # Directly upload a study using the watcher, bypassing the server
+    # Directly upload a local study using the watcher, bypassing the server
     """
     queue.put(('upload', study_name, subject, subject_type, specimen, user, False, False))
     Logger.debug("Study sent to queue directly")
     return 0
 
 
-def upload_sequencing_run_directly(queue, run_name, user, datafiles, reads_type, barcodes_type):
+def upload_sequencing_run_local(queue, run_name, user, datafiles, reads_type, barcodes_type):
     """
-    # Directly upload a sequencing run using the watcher, bypassing the server
+    # Directly upload a local sequencing run using the watcher, bypassing the server
     """
     queue.put(('upload-run', run_name, user, reads_type, barcodes_type, datafiles, False))
     Logger.debug("Sequencing run sent to queue directly")
     return 0
+
+
+def process_sequencing_runs_local(runs_path, queue):
+    """ Uploads sequencing runs included in a dump file """
+    for run in runs_path.glob("*"):
+        # Open directory file and extract paths
+        dir_file = run / fig.SEQUENCING_DIRECTORY_FILE
+        paths = {}
+        with open(dir_file, "r") as f:
+            content = f.read().split('\n')
+            for line in content:
+                if ": " in line:
+                    key, val = line.split(": ")
+                    paths[key] = str(run / val)
+
+        # Evaluate reads type and barcodes type
+        if 'reverse' in paths:
+            reads_type = 'paired_end'
+        else:
+            reads_type = 'single_end'
+        if 'rev_barcodes' in paths:
+            barcodes_type = 'dual_barcodes'
+        else:
+            barcodes_type = 'single_barcodes'
+
+        # Call upload sequencing run
+        result = upload_sequencing_run_local(queue, run.name, fig.REUPLOAD_USER, paths, reads_type, barcodes_type)
+        assert result == 0
+    return 0
+
+
+def process_studies_local(studies_path, queue):
+    """ Uploads studies included in a dump file """
+    for study in studies_path.glob("*"):
+        # Get metadata
+        subject = study / 'subject.tsv'
+        specimen = study / 'specimen.tsv'
+        subject_type = get_subject_type(subject)
+
+        # Get analysis config files
+        i = 0
+        configs = []
+        while (study / f'config_file_{i}.yaml').is_file():
+            configs.append(study / f'config_file_{i}.yaml')
+            i += 1
+
+        # Call upload study
+        result = upload_study_local(queue, study.name, subject, subject_type, specimen, fig.REUPLOAD_USER)
+        assert result == 0
+
+        # Wait for study upload to complete
+        # TODO: This number is arbitrary and could be insufficient for the upload to complete in some cases, fix
```

Comment:
not sure I like this todo, why do we need to sleep the thread here?

Reply:
So the next thing we need to do after this point is upload the analysis config files to the newly created study directory. However, this requires the study directory to have finished being created by the watcher. Usually this just takes a couple of seconds, but with a very large study it could conceivably take quite a while to validate all the cells.

Another solution I was considering was checking every second to see if the folder had been created, but that had some of its own problems. Like if the reupload has been done multiple times, there are already folders that would fit the named criteria of what we're looking for, so how do we know which one is correct? So sleeping for a bit just seemed like the easiest way around. What do you think?

Reply:
Hm, I see what you mean, but that wouldn't prevent the reason we need to sleep the thread in the first place. The wait time comes from the upload itself, not the queue. The queue mainly acts as a way to ensure one upload completes before another begins. I don't think it would be crazy difficult to implement, but I think it's a little outside the scope of this PR. I'd want to be thoroughly sure that the same security that the queue offers would be present in another upload path. Would you be okay with my leaving as is for now, and opening a new issue specifically for that?